Xinfo command

MichalCho opened the pull request:
Hi,

I've noticed that xinfo command is missing so I've started to implement it. 
This is just to start a discussion about how this should be fixed.
Currently I've implemented a single xinfo command that returns a map of <String,Objects>.
This would work for any variant of xinfo (stream,groups, consumers) but it's a bit generic. 

The other idea would be to create 3 xinfo commands, returning something like XInfoStream, XInfoConsumers etc. objects. That would give a better visibility of returned values.
I will be happy to work on it.

MichalCho commented:
@sazzad16 @marcosnils @gkorland could you give me some input?
Thanks

sazzad16 commented:
My initial thought is that `Params` is a better fit for this.

MichalCho commented:
> My initial thought is that `Params` is a better fit for this.

But this will not solve the problem of returning very generic object. 
`jedis.xinfo(myKey, STREAM)` and `jedis.xinfo(myKey, CONSUMERS, myGroup)` would return the same type - just a map. Then user would to need to know the implementation to actually check the returned values example:

`jedis.xinfo(myKey, STREAM).get("first-entry);` - they would need to know that this is a `StreamElement` and do casting in their code.

What I had in mind is to have 3 functions returning different types which expose more clear getters
```
XStreamInfo streamInfo = jedis.xinfo(myKey, STREAM); //Type specific to information about a stream
streamInfo.getFirstEntry(); //returned type is StreamElement, no need for casting;
```

Other option would be to return something like `XInfoElement` that contains 3 fields:
```
XInfoStream
XinfoGroups
XInfoConsumers
```
exposing `getXInfoStream`, `getXInfoGroups` and `XInfoConsumers`.
They would be set when a specific info is requested. That would require if-not-null check.

gkorland commented:
Sorry my mistake... should be draft

MichalCho commented:
> Sorry my mistake... should be draft

No problem, I'm working on the new implementation, I hope I can have some commits ready soon

gkorland commented:
@MichalCho sorry to nudge, but are still working on this PR?

MichalCho commented:
@gkorland no problem. Sorry I got quite busy at the office. I will try to use holiday break now to craft some code. Will you be willing to do some code review?

gkorland commented:
@MichalCho no pressure, I just wanted to check if it's still relevant for you

startjava commented:
@MichalCho   sorry,,,i want know xinfo() method Progress？？ :) :) :D

MichalCho commented:
> @MichalCho sorry,,,i want know xinfo() method Progress？？ :) :) :D

Hi @startjava 
I will try to have some commits ready for review after New Year's Eve.
Are you waiting for this feature? :)

startjava commented:
@MichalCho 
not waiting this feature，But Always waiting this feature  :）:)   :D

MichalCho commented:
Hi,
@gkorland I've just pushed my initial idea how to implement this. It require a lot more testing (I've just created one happy case test case). Do you think you could have a look just at it  to give me feedback what do you think about usability? Then I can continue to work on it after New Year.
Thanks

sazzad16 commented on src/main/java/redis/clients/jedis/BuilderFactory.java:
```diff
@@ -1,13 +1,15 @@
 package redis.clients.jedis;
 
 import java.util.ArrayList;
+import java.util.Collections;
```
Not used.

sazzad16 commented on src/main/java/redis/clients/jedis/BuilderFactory.java:
```diff
@@ -527,6 +529,41 @@ public String toString() {
+
+    @Override
+    public String toString() {
+      return "List<StreamEntry>";
```
Not List.

sazzad16 commented on src/main/java/redis/clients/jedis/BuilderFactory.java:
```diff
@@ -1,13 +1,15 @@
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.stream.Stream;
```
Not used?

sazzad16 commented on src/main/java/redis/clients/jedis/BuilderFactory.java:
```diff
@@ -555,6 +592,143 @@ public String toString() {
+        return null;
+      }
+
+      Map<String,Builder> mappingFunctions = new HashMap<>();
```
You're creating this map during every object creation. Create it once and use rest of the time.

sazzad16 commented on src/main/java/redis/clients/jedis/BuilderFactory.java:
```diff
@@ -555,6 +592,143 @@ public String toString() {
+
+    @Override
+    public String toString() {
+      return "Map<Key,StreamInfoElement>";
```
Doesn't match.

sazzad16 commented on src/main/java/redis/clients/jedis/BuilderFactory.java:
```diff
@@ -555,6 +592,143 @@ public String toString() {
+
+    @Override
+    public String toString() {
+      return "Map<Key,StreamInfoElement>";
```
Doesn't match.

sazzad16 commented on src/main/java/redis/clients/jedis/ShardedJedis.java:
```diff
@@ -14,6 +14,7 @@
 import redis.clients.jedis.params.ZAddParams;
 import redis.clients.jedis.params.ZIncrByParams;
 import redis.clients.jedis.util.Hashing;
+import sun.reflect.generics.reflectiveObjects.NotImplementedException;
```
Not used.

sazzad16 commented on src/main/java/redis/clients/jedis/Jedis.java:
```diff
@@ -3830,6 +3830,48 @@ public long xtrim(final String key, final long maxLen, final boolean approximate
+    resultMap = BuilderFactory.STREAM_INFO.build(streamsEntries);
+
+
+    return resultMap;
```
What if these lines are reduced into 
```
    client.xinfo(key,type);
    List<Object> streamEntries = client.getObjectMultiBulkReply();
    return BuilderFactory.STREAM_INFO.build(streamEntries);
```

Or, further reduced 
```
    client.xinfo(key,type);
    return BuilderFactory.STREAM_INFO.build(client.getObjectMultiBulkReply());
```

sazzad16 commented on src/main/java/redis/clients/jedis/Jedis.java:
```diff
@@ -3830,6 +3830,48 @@ public long xtrim(final String key, final long maxLen, final boolean approximate
+    resultMap = BuilderFactory.STREAM_GROUP_INFO.build(streamsEntries);
+
+
+    return resultMap;
```
See comment on lines +3835 to +3844

sazzad16 commented on src/main/java/redis/clients/jedis/Jedis.java:
```diff
@@ -3830,6 +3830,48 @@ public long xtrim(final String key, final long maxLen, final boolean approximate
+    resultMap = BuilderFactory.STREAM_CONSUMERS_INFO.build(streamsEntries);
+
+
+    return resultMap;
```
See comment on lines +3835 to +3844

sazzad16 commented on src/main/java/redis/clients/jedis/Client.java:
```diff
@@ -1262,5 +1262,29 @@ public void xclaim(String key, String group, String consumername, long minIdleTi
+
+
+    xinfo(SafeEncoder.encode(key),SafeEncoder.encode(StreamInfo.STREAM_INFO));
+
```
Please try to avoid too many blank lines.

sazzad16 commented on src/main/java/redis/clients/jedis/Client.java:
```diff
@@ -1262,5 +1262,29 @@ public void xclaim(String key, String group, String consumername, long minIdleTi
+
+
+    xinfo(SafeEncoder.encode(key),SafeEncoder.encode(StreamGroupInfo.GROUP_INFO));
+
```
Please try to avoid too many blank lines.

sazzad16 commented on src/main/java/redis/clients/jedis/Client.java:
```diff
@@ -1262,5 +1262,29 @@ public void xclaim(String key, String group, String consumername, long minIdleTi
+
+
+    xinfo(SafeEncoder.encode(key),SafeEncoder.encode(group),SafeEncoder.encode(StreamConsumersInfo.CONSUMERS_INFO));
+
```
Please try to avoid too many blank lines.

sazzad16 commented on src/main/java/redis/clients/jedis/Jedis.java:
```diff
@@ -3830,6 +3830,48 @@ public long xtrim(final String key, final long maxLen, final boolean approximate
   }
 
+  @Override
+  public StreamInfo xinfo(String key, StreamInfo.StreamInfoType type) {
```
Put type first. Imitate [Redis order](https://redis.io/commands/xinfo) as much as possible.

sazzad16 commented on src/main/java/redis/clients/jedis/Jedis.java:
```diff
@@ -3830,6 +3830,48 @@ public long xtrim(final String key, final long maxLen, final boolean approximate
+  }
+
+  @Override
+  public List<StreamGroupInfo> xinfo(String key, StreamGroupInfo.StreamGroupInfoType type) {
```
See comment on line +3834

sazzad16 commented on src/main/java/redis/clients/jedis/Jedis.java:
```diff
@@ -3830,6 +3830,48 @@ public long xtrim(final String key, final long maxLen, final boolean approximate
+  }
+
+  @Override
+  public List<StreamConsumersInfo> xinfo(String key, String group, StreamConsumersInfo.StreamConsumersInfoType type) {
```
See comment on line +3834

sazzad16 commented on src/main/java/redis/clients/jedis/BinaryClient.java:
```diff
@@ -1455,4 +1455,22 @@ public void xclaim(byte[] key, byte[] groupname, byte[] consumername, long minId
       sendCommand(XCLAIM, arguments.toArray(new byte[arguments.size()][]));
   }
 
+  public void xinfo(byte[] key, byte[] type) {
```
Change order: type, key

sazzad16 commented on src/main/java/redis/clients/jedis/BinaryClient.java:
```diff
@@ -1455,4 +1455,22 @@ public void xclaim(byte[] key, byte[] groupname, byte[] consumername, long minId
+
+  }
+
+  public void xinfo (byte[] key, byte[] group, byte[] type) {
```
Change order: type, key, group

sazzad16 commented on src/main/java/redis/clients/jedis/BinaryClient.java:
```diff
@@ -1455,4 +1455,22 @@ public void xclaim(byte[] key, byte[] groupname, byte[] consumername, long minId
+    arguments.add(type);
+    arguments.add(key);
+    arguments.add(group);
+    sendCommand(XINFO,arguments.toArray(new byte[arguments.size()][]));
```
Simply 
```
    sendCommand(XINFO, type, key, group);
```

sazzad16 commented:
@MichalCho Even though I've put some review comments, I'm liking the idea. Are we complicating it too much? As the [Redis documentation for XINFO](https://redis.io/commands/xinfo) says

> a well behaving client should fetch the whole list, and report it to the user, for example, **as a dictionary data structure**

Perhaps one method (per class) returning a `Map` is enough? @gkorland WDYT?

MichalCho commented:
@sazzad16 thanks a lot for the review. Regarding formatting comments - I was aware of most of them, I just wanted to show something before leaving for holidays :)


> @MichalCho Even though I've put some review comments, I'm liking the idea. Are we complicating it too much? As the [Redis documentation for XINFO](https://redis.io/commands/xinfo) says
> 
> > a well behaving client should fetch the whole list, and report it to the user, for example, **as a dictionary data structure**
> 
> Perhaps one method (per class) returning a `Map` is enough? @gkorland WDYT?

I was thinking about it. Then I had an idea to provide getters for all map members. Otherwise user will have to try to get on Object from the map and he will have to know what kind of object it is. 

something like (if xinfo returns just a map):
`Long streamLenght info = (Long)jedis.xinfo("xadd-stream1", StreamInfo.StreamInfoType.getStreamInfoType()).get("lenght");`

Another issue that I had in mind is - when we decode the answer we can not be fully flexible as we need to know how to decode the answer (to know that "length" is a integer, "last-generated-id" is a string and "first-entry" is a stream entry object).
One solution would be to try to decode first element that is a name of a field (let's say "length") and then try to decode next element using known type (let's say first try to decode as a string, if we fail, then try to decode as stream object, then as a integer etc) until we succeed. If we fail we throw some kind of decoding exception.
I think that would be the only way to be fully transparent and able to react on changes done on redis side. 
I will try to fix the stuff you have commented on and also craft my new approach for decoding an answer from redis. in few days. In the meantime if you have some thought about it, please let me know.

startjava commented:
@MichalCho hi~ what time xinfo method final over? thank you job !

MichalCho commented:
hi @sazzad16!

I've implemented your comments and changed the returned type just to be `Map<String,Object>`

During the weekend I will try to add more test cases and maybe make the code prettier.
Looking forward to any comments from you and @gkorland 
Thanks

PS @startjava it's ongoing, you can check the latest commit

MichalCho commented:
Hi @sazzad16 and @gkorland 

I've added more tests and worked on the review comments. I think it's ready for a final review from you guys.

Please share your thoughts.
Thanks!

sazzad16 commented on src/main/java/redis/clients/jedis/BinaryClient.java:
```diff
@@ -1455,4 +1455,22 @@ public void xclaim(byte[] key, byte[] groupname, byte[] consumername, long minId
+
+    ArrayList<byte[]> arguments = new ArrayList<>(10);
+    arguments.add(type);
+    arguments.add(key);
```
Note needed.

MichalCho replied:
done

sazzad16 commented on src/main/java/redis/clients/jedis/BinaryClient.java:
```diff
@@ -1455,4 +1455,22 @@ public void xclaim(byte[] key, byte[] groupname, byte[] consumername, long minId
+    ArrayList<byte[]> arguments = new ArrayList<>(10);
+    arguments.add(type);
+    arguments.add(key);
+    arguments.add(group);
```
Note needed.

MichalCho replied:
done

sazzad16 commented on src/main/java/redis/clients/jedis/BuilderFactory.java:
```diff
@@ -527,6 +527,41 @@ public String toString() {
+      if (null == data) {
+        return null;
+      }
+      ArrayList<Object> objectList = (ArrayList<Object>) data;
```
Shouldn't it be `List` instead of ArrayList? Under the hood, it is ArrayList but it isn't *guaranteed* to be so.

MichalCho replied:
done

sazzad16 commented on src/main/java/redis/clients/jedis/StreamGroupInfo.java:
```diff
@@ -0,0 +1,61 @@
+  /*private final String name;
+  private final long consumers;
+  private final long pending;
+  private final String lastDeliveredId;*/
```
May be, keep these variables? Along with the map, as well as in other classes.
The variables will ease implementation. The map would help during changes in Redis.
WDYT?

MichalCho replied:
Do you mean to keep them together with getters? So classes provide both `getXYZInfo()` method returning map and getters for already known fields? I like this idea.

sazzad16 replied:
@MichalCho 

> Do you mean to keep them together with getters? So classes provide both getXYZInfo() method returning map and getters for already known fields?

YES.

MichalCho replied:
ok, the getters are there. And constants can be used to access map elements.
Thanks

MichalCho commented on src/test/java/redis/clients/jedis/tests/commands/StreamsCommandsTest.java:
```diff
@@ -306,6 +305,48 @@ public void xpendeing() {
+    List<StreamConsumersInfo> consumersInfos = jedis.xinfo(StreamConsumersInfo.StreamConsumersInfoType.getStreamGroupInfoType(),
+        "xadd-stream1", "G1");
+
+    System.out.println(info.get(0).getGroupInfo().get("name"));
```
This has already been updated

MichalCho commented on src/main/java/redis/clients/jedis/BuilderFactory.java:
```diff
@@ -555,6 +590,202 @@ public String toString() {
+    }
+  };
+
+  public static final Builder<List<StreamGroupInfo>> STREAM_GROUP_INFO = new Builder<List<StreamGroupInfo>>() {
```
done

MichalCho commented:
I've just committed changes according to review comments. 

I've restored getters for map elements, that are known to be present in a reply as per today.
In case some more elements are added in the future they can still be accessed through the map.

Still the question is - how shall we try to decode those unknown elements. As for now we try use a set of decoding functions that we know are needed to decode specific type of answer. Should all decoding function should be added? 
Anyway I feel we are getting closer to finalize that PR

startjava commented:
please speed final commit       I Wait long time------------------ 原始邮件 ------------------
发件人: "MichalCho"<notifications@github.com>
发送时间: 2020年1月14日(星期二) 晚上6:14
收件人: "xetorthio/jedis"<jedis@noreply.github.com>;
抄送: "startjava"<279377921@qq.com>;"Mention"<mention@noreply.github.com>;
主题: Re: [xetorthio/jedis] Xinfo command (#2069)



I've just committed changes according to review comments.
 
I've restored getters for map elements, that are known to be present in a reply as per today.
 In case some more elements are added in the future they can still be accessed through the map.
 
Still the question is - how shall we try to decode those unknown elements. As for now we try use a set of decoding functions that we know are needed to decode specific type of answer. Should all decoding function should be added?
 Anyway I feel we are getting closer to finalize that PR
 
&mdash;
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.

sazzad16 commented on src/main/java/redis/clients/jedis/BuilderFactory.java:
```diff
@@ -514,7 +514,7 @@ public String toString() {
         Map<String, String> map = new HashMap<>(hash.size()/2);
         while(hashIterator.hasNext()) {
-          map.put(SafeEncoder.encode((byte[])hashIterator.next()), SafeEncoder.encode((byte[])hashIterator.next()));
+          map.put(SafeEncoder.encode(hashIterator.next()), SafeEncoder.encode((byte[])hashIterator.next()));
```
Either change none or change both.

MichalCho replied:
done

MichalCho commented:
Hi @gkorland,

sorry for disturbing but do you think you will have some time anytime soon to review the changes? I'm thinking about adding some java docs to make it easier for users to start working with xinfo

startjava commented:
Not done yet?   It's almost two months.  :(

gkorland commented on src/main/java/redis/clients/jedis/StreamInfo.java:
```diff
@@ -0,0 +1,86 @@
+    private static final StreamInfoType streamInfoType = new StreamInfoType();
+    private StreamInfoType() {
+      //Should not be used
+    };
```
`;`?

MichalCho replied:
The initial idea was to use that type just to overload xinfo function. I've made a constructor private to avoid creating this objects by users. Anyway it's not used anymore

gkorland commented:
I'm sorry it took me so long to really review the code, but I think we should rethink the design and really simplified it. i.e. the different `xinfo` methods/commands are practically different methods.
I would have define three methods 

```java
  @Override
  public StreamInfo xinfoStream(String key) {
    client.xinfoStream(key);

    return BuilderFactory.STREAM_INFO.build(client.getObjectMultiBulkReply());

  }

  @Override
  public List<StreamGroupInfo> xinfoGroup(String key) {
    client.xinfoGroup(key);

    return BuilderFactory.STREAM_GROUP_INFO_LIST.build(client.getObjectMultiBulkReply());

  }

  @Override
  public List<StreamConsumersInfo> xinfoConsumers(String key, String group) {
    client.xinfoConsumers(type,key,group);

    return BuilderFactory.STREAM_CONSUMERS_INFO_LIST.build(client.getObjectMultiBulkReply());

  }
```

MichalCho commented:
HI @gkorland,
thanks for your input. Regarding simplifying - I totally agree, it will be simpler and easier solution to use.
I've come up with my initial design just because I thought that we want to keep the same methods' names as one can use in redis cli (one xinfo command with different arguments).
Do you agree @sazzad16 ?
I will refactor asap.

MichalCho commented:
hi @gkorland,

I've implemented your suggestion - splitting xinfo into 3 different methods. I've also added some simple javadocs to make it easier for a user to understand how to use xinfo with jedis.

I had some doubts regarding how I create info objects, for example:

```
if (map!= null) {
     streamInfo = map;
     length = (Long) map.get(LENGTH);  
```
This can fail if length is not present in the map( for example future version of redis will remove this filed). But according to redis documentation:

> Note that you should not rely on the fields exact position, nor on the number of fields, new fields may be added in the future. 

so I guess it's save to assume that fields that exist now will always be there. Nothing is mentioned about removing fields.

sazzad16 commented on src/main/java/redis/clients/jedis/Protocol.java:
```diff
@@ -80,6 +80,9 @@
 
+  public static final String STREAM = "STREAM";
+  public static final String GROUPS = "GROUPS";
+  public static final String CONSUMERS = "CONSUMERS";
```
Prefix these, perhaps with `XINFO_`. Also put these along with `SENTINEL_...`, `CLUSTER_...`, `PUBSUB_...` variables.

MichalCho replied:
done

MichalCho commented on src/main/java/redis/clients/jedis/commands/BinaryJedisCommands.java:
```diff
@@ -356,4 +356,10 @@
+
+  List<byte[]> xinfoGroup (byte[] key);
+
+  List<byte[]> xinfoConsumers (byte[] key, byte[] group);
```
I think we need some aligment here. I can see that some `BinaryJedisCommands` returns Objects (or Collections of Objects) - in some cases using builder factory directly like 
```@Override
  public List<GeoRadiusResponse> georadiusByMember(final byte[] key, final byte[] member, final double radius,
      final GeoUnit unit, final GeoRadiusParam param) {
    checkIsInMultiOrPipeline();
    client.georadiusByMember(key, member, radius, unit, param);
    return BuilderFactory.GEORADIUS_WITH_PARAMS_RESULT.build(client.getObjectMultiBulkReply());
  }
```
some other returns byte arrays or collection of byte arrays etc.
Maybe I missunderstood the purpose of binary client - I thought it was supposed to use just binary data, without trying to decode them. Could you please help me to understand that?

If we should return objects then I would propose following:
``` Object xinfoStream (byte[] key);

  List<Object> xinfoGroup (byte[] key);

  List<Object> xinfoConsumers (byte[] key, byte[] group);
```

sazzad16 replied:
Following would be good enough:

```
StreamInfo xinfoStream (byte[] key);
List<StreamGroupInfo> xinfoGroup (byte[] key);
List<StreamConsumersInfo> xinfoConsumers (byte[] key, byte[] group);
```

MichalCho replied:
I've refactored it that way and added a test for binary usage.

MichalCho commented:
Pinging  @sazzad16 @gkorland :)

gkorland commented on src/test/java/redis/clients/jedis/tests/commands/StreamsCommandsTest.java:
```diff
@@ -306,6 +319,166 @@ public void xpendeing() {
+    List<StreamConsumersInfo> consumersInfo = jedis.xinfoConsumers(STREAM_NAME, G1);
+
+    //Stream info test
+    assertEquals(2L,streamInfo.getStreamInfo().get(LENGTH));
```
Missing LAST_GENERATED_ID

MichalCho replied:
Done in a way that we check if it's not null - it is difficult to know what last-generated should be

gkorland replied:
shouldn't it be id2?

MichalCho replied:
you are right. I've missed that we have that info already.
Maybe then I should change `getLastGeneratedId` method to return `StreamEntryID` instead of `String`?

gkorland commented on src/test/java/redis/clients/jedis/tests/commands/StreamsCommandsTest.java:
```diff
@@ -306,6 +319,166 @@ public void xpendeing() {
+    assertEquals(2,streamInfo.getLength());
+    assertEquals(1,streamInfo.getRadixTreeKeys());
+    assertEquals(2,streamInfo.getRadixTreeNodes());
+    assertEquals(0,streamInfo.getGroups());
```
Missing getLastGeneratedId()

MichalCho replied:
Done in a way that we check if it's not null - it is difficult to know what last-generated should be

gkorland replied:
shouldn't it be id2?

MichalCho replied:
you are right. I've missed that we have that info already.
Maybe then I should change `getLastGeneratedId` method to return `StreamEntryID` instead of `String`?

gkorland commented on src/main/java/redis/clients/jedis/Protocol.java:
```diff
@@ -73,6 +73,10 @@
   public static final String PUBSUB_NUMSUB = "numsub";
   public static final String PUBSUB_NUM_PAT = "numpat";
 
+  public static final String XINFO_STREAM = "STREAM";
```
Should be added to  `enum Keyword`

MichalCho replied:
done

gkorland replied:
:+1:

gkorland commented on src/main/java/redis/clients/jedis/StreamInfo.java:
```diff
@@ -0,0 +1,89 @@
+
+  private final long length;
+  private final long radixTreeKeys;
+
```
remove white space

MichalCho replied:
done

gkorland commented on src/main/java/redis/clients/jedis/StreamInfo.java:
```diff
@@ -0,0 +1,89 @@
+     firstEntry = (StreamEntry) map.get(FIRST_ENTRY);
+     lastEntry = (StreamEntry) map.get(LAST_ENTRY);
+
+    } else throw  new IllegalArgumentException("InfoMap can not be null");
```
If we are throwing runtime exception anyway when map is null isn't it better to just let it throw NullPointerException?

MichalCho replied:
Done. My initial thought was that it is not nice to see NullPointers

gkorland commented on src/main/java/redis/clients/jedis/StreamGroupInfo.java:
```diff
@@ -0,0 +1,65 @@
+     pending = (long) map.get(PENDING);
+     lastDeliveredId = (String) map.get(LAST_DELIVERED);
+
+    } else throw new IllegalArgumentException();
```
If we are throwing runtime exception anyway when map is null isn't it better to just let it throw NullPointerException?

MichalCho replied:
Done. My initial thought was that it is not nice to see NullPointers

gkorland commented on src/main/java/redis/clients/jedis/StreamConsumersInfo.java:
```diff
@@ -0,0 +1,60 @@
+      pending = (long) map.get(PENDING);
+
+
+    } else throw new IllegalArgumentException();
```
If we are throwing runtime exception anyway when map is null isn't it better to just let it throw NullPointerException?

MichalCho replied:
Done. My initial thought was that it is not nice to see NullPointers

gkorland commented:
Tests are failing https://travis-ci.org/xetorthio/jedis/jobs/648852944

MichalCho commented:
> Tests are failing https://travis-ci.org/xetorthio/jedis/jobs/648852944

I'm on it

gkorland commented:
@MichalCho sounds OK just sleep for 1ms

MichalCho commented:
I think we are done here

sazzad16 commented:
Merged into master. Excellent job @MichalCho

MichalCho commented:
Thanks,
@gkorland @sazzad16 do we have a list of features/ issues to be implemented/fixed? I would like to contribute more?

gkorland commented:
@MichalCho thanks for the great help!
You can see here the tasks marked for the next versions https://github.com/xetorthio/jedis/milestones/
